zebra-chain/src/note_encryption.rs

```diff
@@ -0,0 +1,90 @@
+//! Note encryption types.
+
+use std::{
+    convert::{AsRef, From},
+    fmt,
+};
+
+mod sapling;
+mod sprout;
+
+/// A 512-byte _Memo_ field associated with a note, as described in
+/// [protocol specification §5.5][ps].
+///
+/// The usage of the memo field is by agreement between the sender and
+/// recipient of the note.
+///
+/// [ps]: https://zips.z.cash/protocol/protocol.pdf#spendencoding
+#[derive(Clone, Copy)]
+pub struct Memo([u8; 512]);
+
+impl<T: AsRef<[u8]>> From<T> for Memo {
+    fn from(input: T) -> Self {
+        let input_bytes: &[u8] = input.as_ref();
+
+        let mut full_bytes = [0; 512];
+
+        if input_bytes.len() < 512 {
+            full_bytes[0..input_bytes.len()].copy_from_slice(input_bytes);
+            Memo(full_bytes)
+        } else if input_bytes.len() == 512 {
+            full_bytes[..].copy_from_slice(input_bytes);
+            Memo(full_bytes)
+        } else {
+            // Because this is a From impl, we truncate your input
+            // rather than return any Memo at all.
+            full_bytes.copy_from_slice(&input_bytes[0..512]);
+            Memo(full_bytes)
+        }
+    }
+}
+
+impl fmt::Debug for Memo {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let output: String;
+
+        // This saves work but if the 'valid utf8 string' is just a
+        // bunch of numbers, it prints them out like
+        // 'Memo("\u{0}\u{0}..")', so. ¯\_(ツ)_/¯
+        match String::from_utf8(self.0.to_vec()) {
```

Comment:
This is great for usability. I'm not sure of a better heuristic to use than "is valid UTF-8?" but the downside of printing unicode control characters sometimes seems less bad than never printing text, so I think your choice is a good one.

One small note is that by using the borrowed rather than owned versions of the functions, you can save 2 allocations by using [`str::from_utf8`](https://doc.rust-lang.org/stable/std/str/fn.from_utf8.html) instead of `String::from_utf8` and `&self.0[..]` instead of `to_vec()`.